src/lv_misc/lv_mem.c

```diff
@@ -200,7 +201,18 @@ void * lv_mem_alloc(size_t size)
     if(alloc != NULL) _lv_memset(alloc, 0xaa, size);
 #endif
 
-    if(alloc == NULL) LV_LOG_WARN("Couldn't allocate memory");
+    if(alloc == NULL) {
+      LV_LOG_WARN("Couldn't allocate memory");
+    }else{
+      #if LV_MEM_CUSTOM == 0
+      /* just a safety check, should always be true */
+      if ((uintptr_t) alloc > (uintptr_t) work_mem) {
+        if ((((uintptr_t) alloc - (uintptr_t) work_mem) + size) > mem_max_size) {
+          mem_max_size = ((uintptr_t) alloc - (uintptr_t) work_mem) + size;
```

Comment:
I think it gives wrong result e.g. in this case:
```c
void * m1 = _lv_mem_alloc(100); //100 bytes are used
void * m2 = _lv_mem_alloc(100);  //200 bytes are used
lv_mem_free(m1);                          //100 bytes are used
void * m3 = _lv_mem_alloc(150);  //250 bytes are used but allocated after m2 becasue can't fit m1
```

So using the last pointer is misleading if there are small gaps.

Reply:
Hi

Just tried, and seems OK to me.

```c
static void Test_MemAllocator(void)
{
  printf("Start State\n");
  DebugUart_LVGLMemMonitor();
  printf("Going to Allocate 100 Bytes\n");
  void *m1 = lv_mem_alloc(100); //100 bytes are used
  printf("Allocated at address: %p\n", m1);
  DebugUart_LVGLMemMonitor();
  printf("Going to Allocate 100 Bytes\n");
  void *m2 = lv_mem_alloc(100);  //200 bytes are used
  printf("Allocated at address: %p\n", m2);
  DebugUart_LVGLMemMonitor();
  printf("Going to free first allocation\n");
  printf("Free allocation at address: %p\n", m1);
  lv_mem_free(m1);               //100 bytes are used
  DebugUart_LVGLMemMonitor();
  printf("Going to Allocate 150 Bytes\n");
  void *m3 = lv_mem_alloc(150);  //250 bytes are used but allocated after m2 becasue can't fit to the place of m1
  printf("Allocated at address: %p\n", m3);
  DebugUart_LVGLMemMonitor();
  printf("Going to Allocate 50 Bytes\n");
  void *m4 = lv_mem_alloc(50);  //50 bytes are used should be allocated in the space free from m1
  printf("Allocated at address: %p\n", m4);
  DebugUart_LVGLMemMonitor();
}
```

Gives the following output

> Start State
> LVGL_MEM -> used:   3076 (  7 %), frag:   1 %, biggest free:  45984, max used:   3164
> Going to Allocate 100 Bytes
> Allocated at address: 0x2001a4ec
> LVGL_MEM -> used:   3180 (  7 %), frag:   1 %, biggest free:  45880, max used:   3268
> Going to Allocate 100 Bytes
> Allocated at address: 0x2001a554
> LVGL_MEM -> used:   3284 (  7 %), frag:   1 %, biggest free:  45776, max used:   3372
> Going to free first allocation
> Free allocation at address: 0x2001a4ec
> LVGL_MEM -> used:   3184 (  7 %), frag:   1 %, biggest free:  45776, max used:   3372
> Going to Allocate 150 Bytes
> Allocated at address: 0x2001a5bc
> LVGL_MEM -> used:   3340 (  7 %), frag:   1 %, biggest free:  45620, max used:   3528
> Going to Allocate 50 Bytes
> Allocated at address: 0x2001a4ec
> LVGL_MEM -> used:   3396 (  7 %), frag:   1 %, biggest free:  45620, max used:   3528

We started with a maximum used of 3164, increase to 3268 after first 100 bytes allocation, the maximum memory increase by 104.
After the second allocation the maximum used increased from 3268 to 3372, another increase by 104 bytes.
When we remove the first allocation, the maximum memory used stays the same.
When we allocate a new 150 bytes block, the maximum memory used increased from 3372 to 3528, 156 bytes. Seems OK since the code pads the size to be multiple of 4.
When we allocate a 75 byte block, the maximum memory used continues in 3528 since this block is allocated in the space free from `m1`.

Seems correct, my intention is to have an indication of what should be the minimum buffer size to accommodate the application, even if in the end we ended up with small gaps, in this case in order to run properly the buffer size would need to be at least `3528` although not all of them are in use because of fragmentation.

Reply:
> Seems correct, my intention is to have an indication of what should be the minimum buffer size to accommodate the application, even if in the end we ended up with small gaps, in this case in order to run properly the buffer size would need to be at least 3528 although not all of them are in use because of fragmentation.

I agree, for this purpose your approach is correct. I meant the goal is to show max used memory reduced by the gaps.